Host endpoint

pchaigno opened the pull request:
This pull request adds the special host endpoint (needed for subsequent host network policies PR). See commit messages for details. As a summary:
1. Consolidate host datapath (`bpf_netdev` and `bpf_hostdev_ingress`) into single file `bpf_host`.
0. Remove `FROM_HOST` macro to be able to compile from Go side.
0. Extract source ID resolution from `handle_ipv{4,6}` to reduce complexity.
0. Add `pkg/maps/callsmap` package for constants of internal calls map.
0. Refactor code with new, shared `createEndpoint` function.
0. Main commit. Add special endpoint for host.
0. Add `BPF_V6_NODEPORT` and `BPF_V6_DIRECT_ROUTING` to allow loading IPv6 addresses as either static data or a macro.
0. Fix restore on up/downgrade paths (needed because of header file rename).

qmonnet commented on contrib/vagrant/start.sh:
```diff
@@ -304,7 +304,7 @@ function write_cilium_cfg() {
     filename="${4}"
 
-    cilium_options=" --debug --pprof --enable-k8s-event-handover --k8s-require-ipv4-pod-cidr --auto-direct-node-routes"
+    cilium_options=" --debug --pprof --enable-k8s-event-handover --k8s-require-ipv4-pod-cidr --auto-direct-node-routes --enable-host-firewall --device=enp0s3"
```
This commit should probably be reordered, the option has not been introduced yet.

pchaigno replied:
I've dropped that commit. I'll probably reintroduce as a proper config parameter for the dev. VM later.

qmonnet commented on contrib/vagrant/start.sh:
```diff
@@ -304,7 +304,7 @@ function write_cilium_cfg() {
     filename="${4}"
 
-    cilium_options=" --debug --pprof --enable-k8s-event-handover --k8s-require-ipv4-pod-cidr --auto-direct-node-routes"
+    cilium_options=" --debug --pprof --enable-k8s-event-handover --k8s-require-ipv4-pod-cidr --auto-direct-node-routes --enable-host-firewall --device=enp0s3"
```
This commit should probably be reordered, the option has not been introduced yet.

qmonnet commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -57,6 +57,7 @@ cilium-agent [flags]
       --enable-endpoint-routes                        Use per endpoint routes instead of routing via cilium_host
       --enable-external-ips                           Enable k8s service externalIPs feature (requires enabling enable-node-port) (default true)
       --enable-health-checking                        Enable connectivity health checking (default true)
+      --enable-host-firewall                          Enable host network policies (default deny)
```
For `--enable-something`, I would expect values to be `false`, instead of `deny`? It is not clear what `deny` means for that option.

pchaigno replied:
I actually wanted to highlight that it will implement default-deny policies :laughing: 
I've removed it. We'll need to highlight that in the documentation anyway.

aanm replied:
Does not seem to be resolved in this review.

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -611,10 +607,12 @@ static __always_inline int do_netdev_encrypt(struct __ctx_buff *ctx, __u16 proto
 	__u32 __maybe_unused identity = 0;
-	int ret;
+	int trace = TRACE_FROM_HOST, ret;
+	bool from_proxy;
```
Nit: you can probably leave those definitions under the `if (from_host)` scope.

pchaigno replied:
That seems inconsistent with our usual pattern, no? We always declare variables at the top of the functions.

qmonnet replied:
Ideally I think, at the top of the block in which they are scoped, which is the function most of the time. No strong opinion anyway.

pchaigno replied:
I think #11205 clarifies that; I'll keep declarations in their original scopes.

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -691,15 +685,31 @@ int from_netdev(struct __ctx_buff *ctx)
+		ret = snat_process(ctx, BPF_PKT_DIR);
+		if (ret != CTX_ACT_OK) {
+			return ret;
+		}
```
Nit: curly braces

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -611,10 +607,12 @@ static __always_inline int do_netdev_encrypt(struct __ctx_buff *ctx, __u16 proto
 
-static __always_inline int do_netdev(struct __ctx_buff *ctx, __u16 proto)
+static __always_inline int
+do_netdev_(struct __ctx_buff *ctx, __u16 proto, bool from_host)
```
Is this change of name (`do_netdev_`) intentional? If so, can't we find something else than a trailing underscore? If nothing else, I'd rather have it as a prefix instead.

qmonnet commented on bpf/lib/common.h:
```diff
@@ -448,6 +448,7 @@ enum {
 	CB_SRC_LABEL,
 #define	CB_SVC_PORT		CB_SRC_LABEL	/* Alias, non-overlapping */
 	CB_IFINDEX,
+#define CB_FROM_HOST		CB_IFINDEX	/* Alias, non-overlapping */
```
Tab after `#define` in this block

qmonnet commented on bpf/Makefile:
```diff
@@ -104,16 +104,28 @@ bpf_overlay.o: bpf_overlay.ll
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_MASQUERADE \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_IPSEC:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_MASQUERADE \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_DSR \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID
```
We'll have to find something cleaner to generate those, one day. It's becoming difficult to parse :/.

pchaigno replied:
Yep :disappointed: 
Too bad Coccinelle can't match on preprocessor directives... We could probably do a half-good job by grepping over `bpf_*` files and their includes though.

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -256,11 +262,16 @@ int tail_handle_ipv6(struct __ctx_buff *ctx)
-resolve_srcid_ipv4(struct __ctx_buff *ctx, struct iphdr *ip4,
-		   __u32 srcid_from_proxy, bool from_host)
+resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
+		   bool from_host __maybe_unused)
```
Nit: Should the `__maybe_unused` be added in the commit that introduced `from_host`?

pchaigno replied:
We actually don't need this and it can be removed.

qmonnet commented on pkg/endpoint/id/id.go:
```diff
@@ -59,6 +59,11 @@ const (
 	PodNamePrefix PrefixType = "pod-name"
 
+	// NodeNamePrefix is used to address an endpoint via the Kubernetes node
+	// name. This addressing only works if the endpoint represents as
```
s/as/a/

qmonnet commented on pkg/policy/api/rule.go:
```diff
@@ -34,8 +37,16 @@ import (
+	EndpointSelector EndpointSelector `json:"endpointSelector,omitempty"`
+
+	// EndpointSelector selects all endpoints which should be subject to
+	// this rule. EndpointSelector and NodeSelector cannot be both empty.
```
Should you mention here that the two selectors are mutually exclusive?

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -322,11 +354,178 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
+				     &policy_match_type);
+
+	/* Reply traffic and related are allowed regardless of policy verdict. */
+	if (ret != CT_REPLY && ret != CT_RELATED && verdict < 0) {
```
Could we exit early and avoid the call to `policy_can_egress4()` if we have `ret == CT_REPLY || ret == CT_RELATED`? (Same remark for ingress)

pchaigno replied:
I need to add L7-proxy support later on. We'll need to call `policy_can_{access_ingress,egress}()` for that, regardless of conntrack result.

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -790,7 +1069,45 @@ int to_host(struct __ctx_buff *ctx)
+	ret = CTX_ACT_OK;
+#endif /* ENABLE_HOST_FIREWALL */
+
+	return ret;
```
This makes a lot of duplicated code between `to_host()` and `to_netdev()` :/ Did you consider factorizing this chunk and e.g. passing a bool to choose if we should call ingress or egress policy?

pchaigno replied:
Does it? The structure is the same, but only the switch-case statements are similar, and even there, the `srcID` retrieval is different.

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -1020,13 +1223,15 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 # ifdef ENABLE_IPV4
 	case bpf_htons(ETH_P_IP):
-		// to-netdev is attached on the egress path of the native
+		// to-netdev is attached to the egress path of the native
```
Nit: move to previous commit

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -130,6 +153,173 @@ resolve_srcid_ipv6(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
+	return CTX_ACT_OK;
+}
+#endif /* ENABLE_HOST_FIREWALL */
+
```
(Same remarks as for IPv4)

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -682,6 +765,17 @@ ipv4_host_policy_ingress(struct __ctx_buff *ctx, __u32 *srcID) {
+	}
+
+
+
```
Three blank lines??

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -682,6 +765,17 @@ ipv4_host_policy_ingress(struct __ctx_buff *ctx, __u32 *srcID) {
 
+	if (unlikely(ret == CT_REPLY && ct_state.rev_nat_index &&
+	    	     !ct_state.loopback)) {
+		ret2 = lb4_rev_nat(ctx, l3_off, l4_off, &csum_off,
```
Nit: s/`ret2`/`ret_nat`/?

qmonnet commented on bpf/bpf_netdev.c:
```diff
@@ -203,6 +203,9 @@ ipv6_host_policy_egress(struct __ctx_buff *ctx, __u32 srcID, __u32 *dstID) {
 				return ret;
 		}
+
+		/* 148991 insns before, 114907 after */
```
Nit: Those numbers will change with LLMV version or future code edits, not sure if it might be confusing... Use a approximate percentage of instructions saved instead?

pchaigno replied:
The idea was simply to convey that these calls to `relax_verifier()` are actually useful to pass the verifier. Absolute numbers show that we get below the 128k 4.14 limit (although this was measured on 4.9 so the numbers are even lower on 4.14).

I probably need to revisit these calls as a followup, to ensure they're still needed. I have extended the comment a bit in the meantime.

qmonnet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1469,6 +1471,42 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
+	if option.Config.EnableHostFirewall && egressLBEnabled {
+		// We can probe the verifier's complexity limit (128k here) indirectly
+		// based on what kernel helpers are available since both were changed
+		// in the same release, v4.14.
```
Ugh we really need to add probes to bpftool! :/

qmonnet commented on examples/policies/l3/host/host-policy.json:
```diff
@@ -0,0 +1,41 @@
+[
+  {
+    "nodeSelector": {"matchLabels": {"doh":"no"}},
```
Do we have a way not to put the `matchLabels` in the file, or are they mandatory?

pchaigno replied:
Not sure if that's what you're asking, but if you want to match all nodes regardless of labels, you can do:
```json
    "nodeSelector": {}
```

qmonnet replied:
The idea was that if we can remove the `"doh":"no"` from that example, maybe we should. Not a big concern, but this is the kind of example file that users will tend to joyfully copy-paste from, and we'll end up with plenty of `"doh":"no"` in the wild :)

pchaigno replied:
In case that wasn't clear, "doh" here is not Homer Simpson's expression but short for DNS-over-HTTPS :smiley:  This policy whitelists a couple needed ports for Cilium, then whitelists DNS, but not DOH. That was the point of my short host firewall demo a while back.
I will likely update this example and incorporate it in a getting started guide once I write the documentation followup.

qmonnet reviewed (commented):
Series looks great, thanks!

Some nits and remarks, mostly low-level.
Should we maybe add something to the `Documentation/`? `Documentation/policy/intro.rst` and `Documentation/policy/language.rst` look like good candidates.

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -244,29 +506,70 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
+__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_FROM_HOST)
+int tail_handle_ipv6_from_host(struct __ctx_buff *ctx __maybe_unused)
+{
+	return tail_handle_ipv6(ctx, true);
```
Cool, I was looking at making a similar change for bpf tproxy. :+1:

joestringer commented on bpf/Makefile:
```diff
@@ -104,17 +104,21 @@ bpf_overlay.o: bpf_overlay.ll
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_HOST_FIREWALL \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_HOST_FIREWALL
```
We seemed to lose the `-DLB_L3:-DLB_L4` commit here as part of the main host endpoint commit, but I didn't quite follow why.

joestringer replied:
Maybe be cause logically we won't apply LB there as LB from the host should be done at socket layer and LB from outside should be done at native device before we hit this code..? Either way this was not called out in the commit.

pchaigno replied:
I had a second look at this one. I removed `-DLB_L3:-DLB_L4` because `bpf_netdev` (renamed to `bpf_host`) is now using `lxc_config.h` (renamed to `ep_config.h`) and `lxc_config.h` has `LB_L3` and `LB_L4` always defined. So we don't need to define them here. If we do we'll get an error anyway.

joestringer commented on bpf/init.sh:
```diff
@@ -631,15 +618,9 @@ else
 fi
 
-# bpf_host.o requires to see an updated node_config.h which includes ENCAP_IFINDEX
+COPTS="-DNOT_HOST_ENDPOINT=1 -DSECLABEL=${ID_WORLD}"
```
nit: It's cognitively easier to read `-DHOST_ENDPOINT` and flip that than `-DNOT_HOST_ENDPOINT` and then follow the else case with double negative, ie "not not host endpoint"

pchaigno replied:
If I make it `HOST_ENDPOINT` then I would have to set it from Go, but we load everything through header files there. I'll switch to `POD_ENDPOINT` instead.

joestringer commented on bpf/init.sh:
```diff
@@ -631,15 +618,9 @@ else
-bpf_load $HOST_DEV1 "$COPTS" "egress" bpf_netdev.c bpf_host.o from-netdev $CALLS_MAP
-bpf_load $HOST_DEV1 "" "ingress" bpf_hostdev_ingress.c bpf_hostdev_ingress.o to-host $CALLS_MAP
-bpf_load $HOST_DEV2 "" "ingress" bpf_hostdev_ingress.c bpf_hostdev_ingress.o to-host $CALLS_MAP
+bpf_load $HOST_DEV2 "$COPTS" "ingress" bpf_netdev.c bpf_hostdev_ingress.o to-host $CALLS_MAP
```
We're no longer adding this program to the $HOST_DEV1 ingress, is that intentional?

joestringer replied:
(Overall I found there were a various changes in this particular commit that were not explicitly called out, so I found it a bit difficult to follow which changes were intentional and which were side-effects of reworking the endpoint load stuff into Golang).

pchaigno replied:
> We're no longer adding this program to the $HOST_DEV1 ingress, is that intentional?

You scared me. It is still loaded :-)
`bpf_hostdev_ingress.c` is incorporated in `bpf_netdev.c` and all four attachments to `$HOST_DEV1` (ingress and egress of `cilium_host` and the native device) are performed in `pkg/datapath/loader/loader.go:reloadHostDatapath()`.

> (Overall I found there were a various changes in this particular commit that were not explicitly called out, so I found it a bit difficult to follow which changes were intentional and which were side-effects of reworking the endpoint load stuff into Golang).

I will try to extract all I can from the "Endpoint for host" commit into separate commits. Hopefully that should help clarify some of the changes.

joestringer replied:
Ah, apologies for the scare! I would've been just scanning through and taking notes as I went through and didn't circle back to resolve this issue as "the equivalent code moved elsewhere".

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1251,6 +1269,14 @@ func runDaemon() {
+		log.Info("Creating host endpoint")
+		if err := d.endpointManager.AddHostEndpoint(d.ctx, d, d.l7Proxy, d.identityAllocator,
+			"Create host endpoint", nodeTypes.GetName()); err != nil {
+			log.WithError(err).Fatal("Unable to create host endpoint")
```
What can a user do if they hit this? Is it possible that they unluckily already have endpoint 1 allocated in an existing setup and will randomly hit this error during upgrade?

pchaigno replied:
That's something I wanted to bring up at one of the next meetings. Yes, if someone has an endpoint with ID 1, the upgrade will fail and we need to fix it.

I used 1 for convenience (same as security ID), but I can easily change to something >4k (our current max.). However, the following code comment made me think that no endpoint ID was actually safe:
https://github.com/cilium/cilium/blob/a4867b8af39085220aa2ecda3ced6573b9a43c72/pkg/endpointmanager/idallocator/allocator.go#L56-L58
The max. ID of 4k was introduced in Cilium 1.5. Could someone still be using an endpoint ID  >4k? I would really prefer to have a reserved static ID; I initially had a dynamic ID for the host endpoint, but that makes the code even more convoluted.

Apart from that ID issue, I would consider other cases of `Unable to create host endpoint` as bugs we need to fix.

aanm replied:
Is the ID of an endpoint the only way we can determine if it's host? Why can't it have random IDs and we would determine if it is a host based on the labels associated with it? 

The same way we do it today for the health endpoint.
```
356        Disabled           Disabled          4          reserved:health                                                     10.128.0.76    ready
```

pchaigno replied:
I took a second look at using dynamic endpoint IDs for the host endpoint, and I'm not sure why I abandoned that idea last time. It works fine and removes the upgrade issue (see :point_up:).

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -42,6 +42,11 @@ const (
+	symbolFromHostEp1 = "from-netdev"
+	symbolToHostEp1   = "to-netdev"
+	symbolFromHostEp2 = "from-host"
+	symbolToHostEp2   = "to-host"
```
I'm confused about the four symbol sections here.

One high-level source of the confusion is that `bpf_netdev` has historically handled both the traffic towards Cilium on `cilium_host` dev egress TC qdisc and potentially a native device ingress TC qdisc. Then we had `bpf_hostdev_ingress.c` which was clearly about ingress towards the stack (latest code of which handles encryption + proxy stuff), which in this commit is now folded into the same file. 

If we however ignore native devices for now and focus on the cilium host devices, we have a veth pair, so two devices. Each of which has ingress and egress TC qdisc attachment points. Ingress from either side goes "to the host" and egress from either side goes "to the netdev" (where we may run BPF logic to hijack and send traffic elsewhere, or we may allow the traffic through which then results in the traffic passing to the program on ingress on the other side. This is most commonly used for handing traffic from the host to the proxy).

Our current `bpf/init.sh` script does this:

```
bpf_load $HOST_DEV1 "$COPTS" "egress" bpf_netdev.c bpf_host.o from-host $CALLS_MAP
bpf_load $HOST_DEV1 "" "ingress" bpf_hostdev_ingress.c bpf_hostdev_ingress.o to-host $CALLS_MAP
bpf_load $HOST_DEV2 "" "ingress" bpf_hostdev_ingress.c bpf_hostdev_ingress.o to-host $CALLS_MAP
```

So, we only attach to egress of one of the veth devices (My guess would be `cilium_host`) and that's where we do the handling for traffic from the stack. That's the interface where we would configure routes to direct traffic towards the Cilium managed network.

At a glance I might guess that we never receive traffic on TC egress an the other side of the veth pair; though it's also possible we actually have some hack somewhere in the datapath that explicitly redirects to the egress side of that device just as a way to reinject traffic back into the system (though I can't think off hand of any such examples or why it would be interesting to do so).

Regarding the two ingress programs, they're the same, and we recompile them unnecessarily twice in the current implementation. So overall I see 2 programs that should attach at 3/4 attachment points between them assuming that we don't change the current implementation.

I'm only about 3 commits into the PR right now so maybe I'm missing future context as to why this makes sense. I guess I'll put down my assumptions and background and continue on :-)

pchaigno replied:
I'm not sure what your point is here to be honest :worried: The behavior you're describing should be preserved by this pull request.

So far, I have only considered `cilium_host` and the native device as part of the host endpoint. But maybe it would make things more consistent/readable if I also move the `cilium_net` (i.e., `$HOST_DEV2`) attachment to the Go side.

> I'm only about 3 commits into the PR right now so maybe I'm missing future context as to why this makes sense. I guess I'll put down my assumptions and background and continue on :-)

Don't :-) Subsequent commits introduce the actual policy enforcement and are fairly independent. If things are unclear here, then I need to rework it.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -99,11 +104,57 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+			// this log message should only represent failures with respect to
+			// loading the program.
+			if ctx.Err() == nil {
+				scopedLog.WithError(err).Warn("JoinEP: Failed to load program")
```
I believe that the CI will complain on any "JoinEP" message. It would be useful for this message to be unique compared to the other endpoint ones since that could help narrow down where we focus if there's a failure.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -99,11 +104,57 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+	if option.Config.Device != "undefined" {
+		interfaceNames = append(interfaceNames, option.Config.Device)
+		if option.Config.EnableNodePort || option.Config.EnableHostFirewall ||
+			!option.Config.InstallIptRules && option.Config.Masquerade {
```
Given @brb 's outstanding PR I think it'd be safe to drop this line.

```suggestion
		if option.Config.EnableNodePort || option.Config.EnableHostFirewall {
```

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -424,6 +429,50 @@ func NewEndpointWithState(owner regeneration.Owner, proxy EndpointProxy, allocat
 	return ep
 }
 
+func createEndpoint(owner regeneration.Owner, proxy EndpointProxy, allocator cache.IdentityAllocator, id uint16, ifName string) *Endpoint {
```
Did you consider the overlap with the `NewEndpointWithState` function immediately above this function and whether there are trivial changes we could make for more code sharing here?

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -424,6 +429,50 @@ func NewEndpointWithState(owner regeneration.Owner, proxy EndpointProxy, allocat
+func CreateHostEndpoint(owner regeneration.Owner, proxy EndpointProxy, allocator cache.IdentityAllocator) *Endpoint {
+	ep := createEndpoint(owner, proxy, allocator, idallocator.HostEndpointID, option.Config.HostDevice)
+	ep.mac = []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x1}
+	ep.nodeMAC = []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x1}
```
Will these be properly set later on? These values seem a bit suspicious.

pchaigno replied:
~~These are not actually used and shouldn't be set.~~ Scratch that, they're not used because of a bug I need to fix.

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -460,6 +464,35 @@ func (mgr *EndpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
+	// Give the endpoint a security identity
+	newCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	ep.UpdateLabels(newCtx, epLabels, nil, true)
```
Do we need to put this logic into a controller that runs regularly via `pkg/controller` so that if the user updates the node labels, we will recognize and apply the changes? (Or are we listening for node events and handling that case from events somehow?)

pchaigno replied:
There is a separate commit, later in the series (`pkg/identity: Watch and update labels for the host`), that adds a k8s watcher (only for the local node) for node label updates.

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -460,6 +464,35 @@ func (mgr *EndpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
+	}
+
+	// Give the endpoint a security identity
+	newCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
```
I'm trying to wrap my head around what this 30 second timeout will do if hit and what will happen. Seems like potentially just a silent ignoring of the label changes..?

Maybe we should check the `newCtx` after we run `ep.UpdateLabels()` to ensure that the context deadline was not exceeded (and log a warning or something, I guess?)

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -460,6 +464,35 @@ func (mgr *EndpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
+			log.WithError(err).Warning("Kubernetes node resource representing own node is not available, cannot set Labels")
+		} else {
+			newLabels := labels.Map2Labels(k8sNode.GetLabels(), labels.LabelSourceK8s)
+			newIdtyLabels, _ := labelsfilter.Filter(newLabels)
```
Not sure whether we need special handling for node labels, eg whether the pod labels filter will be appropriate for filtering node labels. Can likely be followed up separately. /cc @aanm

aanm replied:
@joestringer If I understood your question, I think Paul added a k8s watcher for nodes to handle this

joestringer replied:
My concern was that we have `pkg/labelsfilter/` which was built for filtering labels that may apply to pods. Now that we will filter nodes labels as well, there could be additional prefixes we need to ignore / respect.

pchaigno replied:
Looking at `pkg/labelsfilter/filter.go`, I think we do need to update the filtering logic. Maybe updating the list of expressions would be enough though? Unless we need separate lists for pods and nodes? https://github.com/cilium/cilium/blob/ebe156fd7ce87cadaf4cacf3b1cfe151bf9a51d4/pkg/labelsfilter/filter.go#L162-L174

aanm replied:
Perhaps it would be good to have a dedicated filter for nodes. These are the current labels I see on my nodes:
```
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/os: linux
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: k8s1
    kubernetes.io/os: linux
```

pchaigno replied:
I've removed this commit from the PR and we'll submit it as a follow up with the required changes for the label filtering.

joestringer reviewed (commented):
Exciting stuff! I only got about 3-4 commits in as it's a pretty big PR, will come back later.

Is there any chance we can further split this PR? The Github UI is painfully slow to interact with for this PR in its current state.

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -1087,6 +1090,18 @@ func initEnv(cmd *cobra.Command) {
+			log.WithField(logfields.Interface, device).
+				Info("Using auto-derived device for host firewall")
+			option.Config.Device = device
+		}
```
Nit: this block can be de-dented as it can just fall-through from the if statement

christarazi commented on pkg/datapath/linux/config/config.go:
```diff
@@ -438,6 +441,12 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 	}
 
+	if e.IsHost() {
+		if option.Config.EnableNodePort {
```
Nit: can we combine these two conditions together?

pchaigno replied:
This code is extended in a later commit into:
```go
if e.IsHost() {
	fmt.Fprintf(fw, "#define HOST_EP_ID %d\n", idallocator.HostEndpointID)
	if option.Config.EnableNodePort {
		fmt.Fprintf(fw, "#define DISABLE_LOOPBACK_LB 1\n")
	}
}
```
So probably not worth combining conditions here.

christarazi reviewed (approved):
Nice work! Looks good aside from a few comments

aanm commented on pkg/policy/rules.go:
```diff
@@ -157,6 +157,10 @@ func (rules ruleSlice) updateEndpointsCaches(ep Endpoint) (bool, error) {
 	endpointSelected := false
 	for _, r := range rules {
+		// NodeSelector can only match nodes, EndpointSelector only pods.
+		if (r.NodeSelector.LabelSelector != nil) != ep.IsHost() {
```
I've seen `r.NodeSelector.LabelSelector != nil` in a couple places that perhaps it would justify having a method for `api.Rule` that would return a boolean if `api.Rule` is matching a node.

pchaigno replied:
I'll refactor with a method unless we change to `*EndpointSelector` (cf. next comment).

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -70,6 +83,63 @@ func NewRule() *Rule {
 }
 
+// MarshalJSON returns the JSON encoding of Rule r. We need to overwrite it to
+// enforce omitempty on the EndpointSelector nested structures.
```
Wouldn't it be simpler to change `NodeSelector EndpointSelector` to be a pointer instead? (the same for `EndpointSelector EndpointSelector` -> `EndpointSelector *EndpointSelector`)

pchaigno replied:
Current approach requires a bit more code, but changing the type to avoid a serialization issue felt somewhat like a hack. Plus, being quite new to Golang, I wasn't too sure I understood the consequences.
Was there a reason for not using a pointer at first? Or does that really not matter and I should just switch it?

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1085,8 +1089,22 @@ func initEnv(cmd *cobra.Command) {
+	if option.Config.EnableHostFirewall && option.Config.Device == "undefined" {
+		device, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
+		if err != nil {
+			msg := "Host firewall's external facing device could not be determined. Use --device to specify."
```
`... "User --%s to specify",  option.Device)`

pchaigno replied:
The option is `--device=` here ;-)

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -460,6 +464,35 @@ func (mgr *EndpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
+
+	if k8s.IsEnabled() {
+		// Retrieve k8s labels.
+		if k8sNode, err := k8s.GetNode(k8s.Client(), nodeName); err != nil {
```
This node will be the node where `cilium-agent` is running, right?

pchaigno replied:
Yes, but `cilium-agent` runs on all nodes managed by Cilium, right?

aanm replied:
Yes, just trying to be sure.

aanm reviewed (changes requested):
it will require these changes:
```patch
diff --git a/pkg/k8s/apis/cilium.io/v2/client/register.go b/pkg/k8s/apis/cilium.io/v2/client/register.go
index 321546a3c..fc4100095 100644
--- a/pkg/k8s/apis/cilium.io/v2/client/register.go
+++ b/pkg/k8s/apis/cilium.io/v2/client/register.go
@@ -1471,8 +1471,13 @@ var (
                        "either specifically allow the connection or one side has to be omitted.\n\n" +
                        "Either ingress, egress, or both can be provided. If both ingress and egress are " +
                        "omitted, the rule has no effect.",
-               Required: []string{
-                       "endpointSelector",
+               OneOf: []apiextensionsv1beta1.JSONSchemaProps{
+                       {
+                               Required: []string{"endpointSelector"},
+                       },
+                       {
+                               Required: []string{"nodeSelector"},
+                       },
                },
                Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
                        "Description": {
@@ -1490,6 +1495,7 @@ var (
                                },
                        },
                        "endpointSelector": EndpointSelector,
+                       "nodeSelector":     EndpointSelector,
                        "ingress": {
                                Description: "Ingress is a list of IngressRule which are enforced at ingress. " +
                                        "If omitted or empty, this rule does not apply at ingress."
```

and a bump in the version specified on https://github.com/cilium/cilium/blob/24c73acbac5e05b2761586652c6974bcbc129282/pkg/k8s/apis/cilium.io/v2/client/register.go#L56-L58

joestringer commented on bpf/lib/common.h:
```diff
@@ -451,6 +453,7 @@ enum {
 	CB_SRC_LABEL,
 #define	CB_SVC_PORT		CB_SRC_LABEL	/* Alias, non-overlapping */
 	CB_IFINDEX,
+#define	CB_FROM_HOST		CB_IFINDEX	/* Alias, non-overlapping */
```
I didn't further review the PR, but while rebasing some of my work to see how it interacts, I noticed that I will need this equivalent bit as an alias for a higher enum as it would conflict with `CB_IFINDEX` in the case where I was going to use it. For context I'm looking at using it to transfer this information across the the tail calls from `ipv[46]_local_delivery()`.

No need to make changes here but I will likely move to shift this CB enum to alias `CB_POLICY` instead (unless you otherwise have reservations on that, then I'll keep searching the higher bits for it, or define a semantically similar bit with a different name and enum.)

pchaigno replied:
I had forgotten that I actually removed `CB_FROM_HOST` in a subsequent commit toward the end of the series to address complexity issues. Instead, I've broken `CILIUM_CALL_IPV{4,6}_FROM_LXC` into `CILIUM_CALL_IPV{4,6}_{FROM_LXC,FROM_HOST}`. Then, in the tail called programs, `from_host` is derived from the above entry points. That makes it known at compile time and allows the compiler to reduce the program size to <4096 once policy enforcement is introduced.

I've merged those two commits (second commit) so that `CB_FROM_HOST` doesn't appear in the PR anymore.

aanm commented on pkg/endpointmanager/idallocator/allocator.go:
```diff
@@ -21,6 +21,10 @@ import (
 
 const (
+	// Reversed IDs are at the end of the 16-bits range to ease any future
+	// extension of the normal ID range.
```
`Reversed` or `Reserved`?

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -334,6 +336,10 @@ func (e *Endpoint) LXCMac() mac.MAC {
 }
 
+func (e *Endpoint) IsHost() bool {
+	return e.ID == idallocator.HostEndpointID
```
Yeah and this function would return true if the endpoint contains the `reserved:host` label

pchaigno replied:
I took a slightly different approach here. Instead of checking labels every time we call `IsHost()`, I introduced a new `isHost` boolean in `Endpoint`. When we create the host endpoint we can just set the boolean; when we restore the host endpoint, we can derive it once from labels.

joestringer replied:
FWIW the labels selection should be derived once and cached assuming we use the standard Cilium structures. (This was a high-impact policy optimization a couple of releases back)

pchaigno replied:
To match against rules yes, but `Endpoint.HasLabels()` just iterates over labels. Anyway, setting and using a boolean also avoided some locking issues on labels :-)

aanm reviewed (commented):
empty comment, discussed offline

joestringer commented on Documentation/_static/cilium_bpf_egress.svg:
```diff
@@ -1,2 +1,2 @@
 <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
-<svg xmlns="http://www.w3.org/2000/svg" style="background-color: rgb(255, 255, 255);" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="1458px" height="934px" viewBox="-0.5 -0.5 1458 934" content="&lt;mxfile modified=&quot;2019-07-30T03:53:14.902Z&quot; host=&quot;www.draw.io&quot; agent=&quot;Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0&quot; version ...
\ No newline at end of file
+<svg xmlns="http://www.w3.org/2000/svg" style="background-color: rgb(255, 255, 255);" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="1458px" height="934px" viewBox="-0.5 -0.5 1458 934" content="&lt;mxfile modified=&quot;2019-07-30T03:53:14.902Z&quot; host=&quot;www.draw.io&quot; agent=&quot;Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0&quot; version ...
```
Thanks for updating the SVGs!

nit: We should update the Cilium version text as well, they still say "v1.6".

joestringer commented on Documentation/_static/cilium_bpf_ingress.svg:
```diff
@@ -1,2 +1,2 @@
 <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
-<svg xmlns="http://www.w3.org/2000/svg" style="background-color: rgb(255, 255, 255);" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="1649px" height="783px" viewBox="-0.5 -0.5 1649 783" content="&lt;mxfile modified=&quot;2019-07-30T03:57:00.605Z&quot; host=&quot;www.draw.io&quot; agent=&quot;Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0&quot; version ...
\ No newline at end of file
+<svg xmlns="http://www.w3.org/2000/svg" style="background-color: rgb(255, 255, 255);" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="1649px" height="783px" viewBox="-0.5 -0.5 1649 783" content="&lt;mxfile modified=&quot;2019-07-30T03:57:00.605Z&quot; host=&quot;www.draw.io&quot; agent=&quot;Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0&quot; version ...
```
nit: looking at this one though it seems that we need to do a proper review before v1.8 given there's some out-of-date bits here like the standalone LB.

pchaigno replied:
Filed #11491 for that.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -36,7 +46,7 @@
 #include "lib/eps.h"
 
-#if defined(FROM_HOST) && (defined(ENABLE_IPV4) || defined(ENABLE_IPV6))
+#if defined(ENABLE_IPV4) || defined(ENABLE_IPV6)
```
nit: Do we need this check any more? One of these two must be true, no?

pchaigno replied:
I can't remove it for now because [we compile test with both IPv4 and IPv6 disabled](https://github.com/cilium/cilium/blob/d613dea5d52494da703e72ef0011e2849f7f19ce/bpf/Makefile#L37). I'm wondering if that makes sense though. If we decide it doesn't, I can remove it in a follow up.

joestringer commented on bpf/init.sh:
```diff
@@ -627,8 +572,8 @@ for iface in $(ip -o -a l | awk '{print $2}' | cut -d: -f1 | cut -d@ -f1 | grep
-        if tc filter show dev "$iface" "$where" | grep -q "bpf_netdev.o"; then
-            echo "Removing bpf_netdev.o from $where of $iface"
+        if tc filter show dev "$iface" "$where" | grep -q "bpf_host.o"; then
+            echo "Removing bpf_host.o from $where of $iface"
```
Q: How do we ensure that the old `bpf_netdev.o` is also removed (eg during upgrade)?

pchaigno replied:
Hm, I had missed this bit of logic. Regarding `bpf_netdev.o` on upgrades it should be enough to just grep for both `bpf_netdev.o` and `bpf_host.o`. I also need to add logic on the Go side to unload programs from the egress of native devices if NodePort is disabled.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -72,25 +81,29 @@ derive_src_id(const union v6addr *node_ip, struct ipv6hdr *ip6, __u32 *identity)
+		ret = derive_src_id(&node_ip, ip6, &src_id);
+		if (IS_ERR(ret))
+			return ret;
+	}
```
:thinking: 

Q: The previous macro did `#ifndef FROM_HOST`, but the new version here is doing the opposite, `if (from_host)`.  Was this intentional?

pchaigno replied:
Not intentional and kinda worrying that it didn't break anything in CI. Will fix.

tgraf replied:
This is the test that could potentially fail:
```
                It("Check connectivity with automatic direct nodes routes", func() {
                        deploymentManager.DeployCilium(map[string]string{
                                "global.tunnel":               "disabled",
                                "global.autoDirectNodeRoutes": "true",
                        }, DeployCiliumOptionsAndDNS)

                        Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
                })
```

Unfortunately, it does not enforce any policies which is why it didn't break. I've filed https://github.com/cilium/cilium/issues/11486

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -402,19 +409,31 @@ func (e *Endpoint) waitForProxyCompletions(proxyWaitGroup *completion.WaitGroup)
 func NewEndpointWithState(owner regeneration.Owner, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, state string) *Endpoint {
+	ep := createEndpoint(owner, proxy, allocator, ID, "")
+	ep.state = state
+	ep.eventQueue = eventqueue.NewEventQueueBuffered(fmt.Sprintf("endpoint-%d", ID), option.Config.EndpointQueueSize)
+
+	ep.UpdateLogger(nil)
+
+	ep.eventQueue.Run()
```
I'm a little surprised to see that these fields are not initialized in the main createEndpoint() function, and AFAICT at least in this PR, they will remain uninitialized. What's the potential impact?

Two extension questions:
1. Relating to `eventQueue`, what is the intention for handling policy updates?
1. Relating to `UpdateLogger`, I presume that this handles `cilium endpoint log ...` CLI. What will happen when you run this with the host endpoint?

(Maybe these questions will be answered in the second PR)

joestringer replied:
Digging a little deeper, I'm a bit surprised that we have a path that skips initializing these today and I'm a little lost as to why this initialization is spread across multiple places in the codebase --- these seem to be handled in `pkg/endpoint/manager.go:Expose()` for at least one of the paths. I guess the intent is to only start handling events from the queue once the endpoint has been exposed, perhaps to simplify endpoint cleanup if another error occurs between instantiating the endpoint object and exposing it via the endpointmanager. Probably we can get away with avoiding handling this for now, but it'd be good to review these paths once we get the initial version of host fw in.

joestringer replied:
Looping back around to answer my own question here, these are handled by `Expose()` for the host endpoint as well in this PR so there's nothing specific to this PR about the question. They will be initialized, it's just a bit of an unusual pattern.

pchaigno replied:
> I'm a bit surprised that we have a path that skips initializing these today and I'm a little lost as to why this initialization is spread across multiple places in the codebase

Yep, that wasn't the easiest to understand code when I started. Plus, I still don't understand the exact purpose of all the fields initialized here.

This PR isn't supposed to change the behaviors you describe though. `createEndpoint()` only does some code refactoring to avoid duplicate code and `CreateHostEndpoint()` should follow the general pattern of other endpoint creations.

tgraf replied:
I agree with @joestringer that we should look at this outside of the context of this PR.

joestringer replied:
> This PR isn't supposed to change the behaviors you describe though. createEndpoint() only does some code refactoring to avoid duplicate code and CreateHostEndpoint() should follow the general pattern of other endpoint creations.

I agree, as far as I can tell this is not changed by this PR.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -375,29 +377,53 @@ func (h *HeaderfileWriter) WriteNetdevConfig(w io.Writer, cfg datapath.DeviceCon
+			}
+		}
+
+		fmt.Fprint(fw, defineUint32("HOST_EP_ID", uint32(e.GetID())))
```
Note: The commit message here still refers to the special endpoint ID 1, which I believe is no longer reserved for this purpose.

pchaigno replied:
If you are referring to `It takes the reserved security ID 1.` then this is still true. The host endpoint will have a dynamic *endpoint ID* and a fixed, reserved *security ID*. The latter is 1 as that's what we use today for the host in policies.

joestringer replied:
Ah right, sorry I misread it. You're completely right.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -375,29 +377,53 @@ func (h *HeaderfileWriter) WriteNetdevConfig(w io.Writer, cfg datapath.DeviceCon
+			fmt.Fprint(fw, defineUint32("DIRECT_ROUTING_DEV_IFINDEX", 1))
+			fmt.Fprint(fw, defineUint32("NATIVE_DEV_IFINDEX", 1))
+			if option.Config.EnableIPv6 {
+				placeholderIPv6 := []byte{1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1}
+				fmt.Fprint(fw, defineIPv6("IPV6_DIRECT_ROUTING", placeholderIPv6))
+				fmt.Fprint(fw, defineIPv6("IPV6_NODEPORT", placeholderIPv6))
+			}
+			if option.Config.EnableIPv4 {
+				placeholderIPv4 := []byte{1, 1, 1, 1}
```
nit: Should we use better fallback IPs here in case this does accidentally trigger some code now or in the future?

Eg using 192.0.2.x/24, not sure what an appropriate IPv6 nonsense / intended unroutable range is.

joestringer replied:
(I noted later on that these get substituted for all 0s which I believe is also somewhat nonsense as an IP so perhaps this point is moot.)

pchaigno replied:
> I noted later on that these get substituted for all 0s which I believe is also somewhat nonsense as an IP

That's correct, and 0 is not valid as a destination IP address. The top comment here should make this clearer though.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -99,10 +109,150 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+	}
+
+	if option.Config.EnableNodePort {
+		// First device from the list is used for direct routing between nodes
```
/cc @brb for when we fix up the daemon options on the configuration of this.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -99,10 +109,150 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+// Since we only want to replace one int symbol, we can nullify string
+// substitutions with:
+//   string_for_endpoint -> string_for_endpoint
+// We cannot simply pass an empty map as the agent would complain that some
+// symbol had no corresponding values.
```
nit: We could probably tweak a version of the function to avoid these complaints in a case like this where we explicitly want the template to be identical except for this one thing. (With a big warning in the comment above that tells developers not to use it if they don't know what they're doing :-) )

pchaigno replied:
Added this to my list of follow ups.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -99,10 +109,150 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+	return hostObj.Write(dstPath, opts, strings)
+}
+
+func (l *Loader) reloadHostDatapath(ctx context.Context, ep datapath.Endpoint, objPath string) error {
```
nit: Is `objPath` here really a template for the host endpoint? Might be easier to read below if this were more clear.

pchaigno replied:
It is not THE template for the host endpoint. It's the object file that will be attached to the host device (i.e., `cilium_host`). That object file was derived from the host endpoint located in `state/template/` as usual. That being said, the principle is the same since we derive patched object files for native devices from the object file for the host device. Not sure if that's enough to address your comment, but I will add a function comment to make this clearer.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -99,10 +109,150 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+func (l *Loader) reloadHostDatapath(ctx context.Context, ep datapath.Endpoint, objPath string) error {
+	symbols := []string{symbolToHostEp, symbolFromHostEp}
+	directions := []string{dirIngress, dirEgress}
+	objPaths := []string{objPath, objPath}
```
nit: We could probably preallocate the array sizes here and for `interfaceNames` using the length of `option.Config.Devices`.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -99,10 +109,150 @@ func removeEndpointRoute(ep datapath.Endpoint, ip net.IPNet) error {
+			if ctx.Err() == nil {
+				scopedLog.WithError(err).Warningf("JoinEP: Failed to load program for host endpoint (%s)", symbol)
+			}
+			return err
```
nit: Did you consider the potential implications of returning on the first error here?

Not sure the full set of cases this might affect, but here's a straw man:

If a user specifies a bad device name in the list of devices, then it looks like this function would successfully load the BPF programs for the base host functionality, as well as any "good" device names in the list prior to the "bad" device name, then fail to load the prog onto the bad device name and any subsequent devices. Though the error would be propagated up as well, not sure how that ends up impacting the higher-layer logic.

(I don't *think* we have a lot to worry about here, more just seeing if you'd considered it)

pchaigno replied:
I will add a sanity check on the attach links (just checking they exist). We already have a similar check before calling `init.sh`.

Nevertheless, our current code would also fail in a similar manner if there's an attach error or if the link disappears between the sanity check and the actual loading (since `init.sh` has `set -e`). I'd rather not fix/change that behavior in this PR, but maybe we should skip any failing device in a follow up PR instead of skipping all devices on any fail.

joestringer commented on pkg/testutils/endpoint.go:
```diff
@@ -61,6 +73,7 @@ func (e *TestEndpoint) GetIdentity() identity.NumericIdentity   { return e.Ident
 func (e *TestEndpoint) GetSecurityIdentity() *identity.Identity { return e.Identity }
 func (e *TestEndpoint) GetNodeMAC() mac.MAC                     { return e.MAC }
 func (e *TestEndpoint) GetOptions() *option.IntOptions          { return e.Opts }
+func (e *TestEndpoint) IsHost() bool                            { return false }
```
nit: So the endpoint returned by `NewTestHostEndpoint()` will return `false` to calls to `IsHost()`?

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -134,13 +134,24 @@ func (e *Endpoint) writeHeaderfile(prefix string) error {
+		if err := os.Symlink(common.CHeaderFileName, oldHeaderPath); err != nil {
+			e.getLogger().Warn("Failed to create C header file symlink")
+		}
+	}
```
What kind of validation have you done to ensure that the implementation will handle the symlink correctly? (Do we think that the standard upgrade test should properly handle this?)

pchaigno replied:
I validated this using the existing K8sUpdates test (doing upgrades and downgrades between v1.7 and master). It took several iterations to get this right so I know it is actually tested. I also checked manually the set of files and symlinks present at different points during the test (e.g., after downgrade, after upgrade).

The logic may not be obvious for all cases (e.g., skip restore of host endpoint on downgrade), so I tried to add comments wherever it makes sense why each piece is needed and when we won't need them anymore. I can add more details in comments if anything remains unclear.

joestringer reviewed (changes requested):
Nice work!

There's a few comments below which I think are important to address or discuss before merge, but they're fairly self-contained to assuming there's a reasonable answer then my overall review would be to approve.
* Reversal of the `FROM_HOST` logic.
* Ensuring the old bpf_netdev.o prog is removed from the host devices.
* Validation that earlier Cilium versions handle the symlink fine.

Several of my other comments are minor nits which would be appropriate to consider and either add to a list of followups or thoughtfully discard :-)

In general I feel as though it'd be nice to abstract the endpoint a little bit more to avoid having these `IsHost()` checks sprinkled all over the place, but this is not a blocker to getting the base functionality in.

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -460,6 +463,28 @@ func (mgr *EndpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
+	// Give the endpoint a security identity
+	newCtx, cancel := context.WithTimeout(ctx, launchTime)
+	defer cancel()
+	ep.UpdateLabels(newCtx, labels.LabelHost, nil, true)
```
Is `labels.LabelHost` defined anywhere?

pchaigno replied:
Yes, in `pkg/labels/labels.go`, next to `labels.LabelHealth`.

tgraf commented on daemon/cmd/daemon_main.go:
```diff
@@ -1288,6 +1288,14 @@ func runDaemon() {
 	restoreComplete := d.initRestore(restoredEndpoints)
 
+	if !d.endpointManager.HostEndpointExists() {
+		log.Info("Creating host endpoint")
```
What is the purpose of this log message?

pchaigno replied:
Simply an indication of where we are in the main initialization (and whether we needed to create host endpoint). We have these sorts of info. messages at [several places in `runDaemon()`](https://github.com/cilium/cilium/blob/f13888d203cc58d9c27346d62d5f085ae936417d/daemon/cmd/daemon_main.go#L1268) and during [the initialization of the health endpoint](https://github.com/cilium/cilium/blob/26308b6304ec9bd1326ac1cdb715ab51cb5b198a/daemon/cmd/health.go#L42) as well.

Happy to remove if we think that one is superfluous.

pchaigno committed:
bpf: Macros to load IPV6_{NODEPORT,DIRECT_ROUTING}

We define IPV6_NODEPORT and IPV6_DIRECT_ROUTING both from Golang (for
bpf_netdev) and init.sh (for bpf_xdp). On Go side it is defined as static
data whose value is patched before loading. In init.sh the values are
given as simple macros. IPv6 addresses defined as static data rely on
BPF_V6(), but BPF_V6() doesn't support macros.

Thus, we define BPF_V6_NODEPORT() and BPF_V6_DIRECT_ROUTING() which will
fallback to BPF_V6 in case of static data and just copy the value
otherwise.

For IPV6_DIRECT_ROUTING, this workaround can be removed once we move
it's definition to node_config.h. For IPV6_NODEPORT, a better, long-term
fix would be to load the bpf_xdp program from Golang.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
Endpoint for host

This commit adds a special Cilium endpoint on each node to represent the
node itself. It takes the reserved security ID 1. The special endpoint is
created by endpoint.CreateLocalNodeEndpoint on daemon startup if that
endpoint wasn't already restored.

When the datapath for the special endpoint is regenerated, it loads
bpf_host.o instead of bpf_lxc.o. bpf_host enforces policies for both
traffic to/from pods and traffic to/from the outside world. It follows
the same template generation path as traditional bpf_lxc files.

ENABLE_EXTRA_HOST_DEV was only defined in init.sh for the from_host path.
It is now defined from the Go side for all paths, and we check whether
from_host is true in the C code instead.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Move host datapaths into bpf_host.c

bpf_netdev.c is renamed to bpf_host and bpf_hostdev_ingress is included,
in preparation for the host endpoint that requires a single object file.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Remove FROM_HOST macro

For the host firewall, we need to have a single bpf_host object file
containing both the from-host and the from-netdev paths, to load it from
the Go side. We therefore turn FROM_HOST into a variable set to true or
false depending on the entry point (i.e., from_host() vs. from_netdev()).

The other option (to avoid the increased code complexity in datapath) is
to let the Go side handle several object files and header files for the
special host endpoint. That second option is likely to just move the
increase code complexity to the Go side.

The CILIUM_CALL_IPV{4,6}_FROM_LXC entry points (and handle_ipv{4,6}) are
broken into CILIUM_CALL_IPV{4,6}_FROM_LXC and
CILIUM_CALL_IPV{4,6}_FROM_HOST based on the value of from_host. This
change allows the compiler to perform dead code elimination on from_host
and reduce the program sizes to <4096.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
pkg/endpoint: Fix up/downgrade paths

Because we changed the C header filenames (from lxc_config.h to
ep_config.h), endpoints don't get properly restored on upgrades and
downgrades.

This commit fixes the up/downgrade paths by 1) creating a lxc_config.h
symlink for new ep_config.h header files and 2) renaming the header file
and creating the symlink when restoring endpoints on an upgrade. On a
downgrade, Cilium will pick up the lxc_config.h symlink.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
pkg/maps: callsmap package for internal calls map

The package contains a single constant definition right now but will
have more with the host endpoint in later commits.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
endpoint: Introduce shared createEndpoint function

This function is also going to be used to create the host endpoint in
subsequent commits.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Extract source ID resolution from handle_ipv{4,6}

handle_ipv{4,6} have the highest complexity as reported by the BPF
verifier. We can extract the source ID resolution from these programs
and execute it before the tail calls. This comes at zero cost in term of
metadata slots since we already use the same metadata slot to transfer the
source identity from the proxy across the tail call.

Signed-off-by: Paul Chaignon <paul@cilium.io>

joestringer commented on bpf/init.sh:
```diff
@@ -627,10 +589,14 @@ for iface in $(ip -o -a l | awk '{print $2}' | cut -d: -f1 | cut -d@ -f1 | grep
 	$found && continue
     for where in ingress egress; do
-        if tc filter show dev "$iface" "$where" | grep -q "bpf_netdev.o"; then
+        if tc filter show dev "$iface" "$where" | grep -qP "bpf_netdev.o"; then
```
nit: It'd be good to add a comment here stating that this old prog name was loaded onto devices in Cilium 1.7 or earlier and this snippet could be removed in v1.9.

joestringer reviewed (approved):
:shipit:

pchaigno committed:
endpoint: Use CIDR prefixes from Daemon for host endpoint

Because the host endpoint relies on bpf_host instead of bpf_lxc, it
needs the CIDR prefixes maintained by the Daemon instead of
endpoint-specific CIDR prefixes. The former are already printed in
netdev_config.h but the host endpoint datapath doesn't use that header
file anymore. Instead, we need to print the appropriate CIDR prefixes
(the Daemon's) into the host endpoint's ep_config.h header file.

We therefore also need to regenerate the host endpoint whenever CIDR
policies change, as we already do for the base programs.

Signed-off-by: Paul Chaignon <paul@cilium.io>
